plugins/parsers/csv/parser.go

```diff
@@ -17,33 +19,98 @@ import (
-	SkipRows          int      `toml:"csv_skip_rows"`
-	TagColumns        []string `toml:"csv_tag_columns"`
-	TimestampColumn   string   `toml:"csv_timestamp_column"`
-	TimestampFormat   string   `toml:"csv_timestamp_format"`
-	Timezone          string   `toml:"csv_timezone"`
-	TrimSpace         bool     `toml:"csv_trim_space"`
-	SkipValues        []string `toml:"csv_skip_values"`
+	ColumnNames        []string `toml:"csv_column_names"`
+	ColumnTypes        []string `toml:"csv_column_types"`
+	Comment            string   `toml:"csv_comment"`
+	Delimiter          string   `toml:"csv_delimiter"`
+	HeaderRowCount     int      `toml:"csv_header_row_count"`
+	MeasurementColumn  string   `toml:"csv_measurement_column"`
+	MetricName         string   `toml:"metric_name"`
+	SkipColumns        int      `toml:"csv_skip_columns"`
+	SkipRows           int      `toml:"csv_skip_rows"`
+	TagColumns         []string `toml:"csv_tag_columns"`
+	TimestampColumn    string   `toml:"csv_timestamp_column"`
+	TimestampFormat    string   `toml:"csv_timestamp_format"`
+	Timezone           string   `toml:"csv_timezone"`
+	TrimSpace          bool     `toml:"csv_trim_space"`
+	SkipValues         []string `toml:"csv_skip_values"`
+	MetadataRows       int      `toml:"csv_metadata_rows"`
+	MetadataSeparators []string `toml:"cvs_metadata_separators"`
+	MetadataTrimSet    string   `toml:"cvs_metadata_trim_set"`
+
+	MetadataSeparatorList MetadataPattern
 
 	gotColumnNames bool
 
-	TimeFunc    func() time.Time
-	DefaultTags map[string]string
+	TimeFunc     func() time.Time
+	DefaultTags  map[string]string
+	MetadataTags map[string]string
 }
 
 // Parser is a CSV parser, you should use NewParser to create a new instance.
 type Parser struct {
 	*Config
 }
 
+type MetadataPattern []string
+
+func (record MetadataPattern) Len() int {
+	return len(record)
+}
+func (record MetadataPattern) Swap(i, j int) {
+	record[i], record[j] = record[j], record[i]
+}
+func (record MetadataPattern) Less(i, j int) bool {
+	// Metadata with longer lengths should be ordered before shorter metadata
+	return len(record[i]) > len(record[j])
+}
+
+func getUniqueMetadataPatterns(patternMap []string) MetadataPattern {
+	result := MetadataPattern{}
+	patternList := map[string]bool{}
+	for _, p := range patternMap {
+		if patternList[p] {
+			// Ignore further, duplicated entries
+			continue
+		}
+		patternList[p] = true
+		result = append(result, p)
+	}
+	return result
+}
+
+func (c *Config) initializeMetadataSeparators() error {
+	// initialize metadata
+	c.MetadataTags = map[string]string{}
+	c.MetadataSeparatorList = []string{}
+
+	if c.MetadataRows > 0 {
+		if len(c.MetadataSeparators) == 0 {
+			return fmt.Errorf("when csv_metadata_rows is defined, " +
+				"cvs_metadata_separators must have at least one valid separator string")
+		}
+		c.MetadataSeparatorList = getUniqueMetadataPatterns(c.MetadataSeparators)
+		sort.Stable(c.MetadataSeparatorList)
+	}
+	return nil
+}
```

Comment:
How about fusing the two and reorganize the code a bit to save some indention level?
```suggestion
func (c *Config) initializeMetadataSeparators() error {
	// initialize metadata
	c.MetadataTags = map[string]string{}
	c.MetadataSeparatorList = []string{}

	if c.MetadataRows == 0 {
		return nil
	}
	if len(c.MetadataSeparators) == 0 {
		return fmt.Errorf("csv_metadata_separator required when specifying csv_metadata_rows")
	}

	patternList := map[string]bool{}
	for _, p := range patternMap {
		if patternList[p] {
			// Ignore further, duplicated entries
			continue
		}
		patternList[p] = true
		c.MetadataSeparatorList = append(c.MetadataSeparatorList, p)
	}
	sort.Stable(c.MetadataSeparatorList)
	return nil
}
```